eth/filters/api.go

```diff
@@ -246,31 +251,67 @@ func (api *FilterAPI) NewHeads(ctx context.Context) (*rpc.Subscription, error) {
 	return rpcSub, nil
 }
 
-// Logs creates a subscription that fires for all new log that match the given filter criteria.
+// notifier is used for broadcasting data(eg: logs) to rpc receivers
+// used in unit testing.
+type notifier interface {
+	Notify(id rpc.ID, data interface{}) error
+	Closed() <-chan interface{}
+}
+
+// Logs creates a subscription that fires for all historical
+// and new logs that match the given filter criteria.
 func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subscription, error) {
 	notifier, supported := rpc.NotifierFromContext(ctx)
 	if !supported {
 		return &rpc.Subscription{}, rpc.ErrNotificationsUnsupported
 	}
+	rpcSub := notifier.CreateSubscription()
+	err := api.logs(ctx, notifier, rpcSub, crit)
+	return rpcSub, err
+}
 
-	var (
-		rpcSub      = notifier.CreateSubscription()
-		matchedLogs = make(chan []*types.Log)
-	)
+// logs is the inner implementation of logs subscription.
+// The following criteria are valid:
+// * from: nil, to: nil -> yield live logs.
+// * from: blockNum | safe | finalized, to: nil -> historical beginning at `from` to head, then live logs.
+// * Every other case should fail with an error.
+func (api *FilterAPI) logs(ctx context.Context, notifier notifier, rpcSub *rpc.Subscription, crit FilterCriteria) error {
+	if crit.ToBlock != nil {
+		return errInvalidToBlock
+	}
+	if crit.FromBlock == nil {
+		return api.liveLogs(notifier, rpcSub, crit)
```

Comment:
on input, you take a `ctx`. But the ctx is lost, afaict you don't check for ctx timeouts or cancel. Shouldn't you? What is the lifecycle here?

Reply:
As the `ctx` will be canceled when the subscription goroutine returns, so we instantiated a new background context to control the workflow. 

The lifecycle is as below:

1. do historical logs fetch and push;
2. do live logs push;
3. terminate if the subscription was canceled or the push channel was broken;